modules/core/arrow-test/src/main/kotlin/arrow/test/laws/ZipLaws.kt

```diff
@@ -0,0 +1,190 @@
+      val rs = (x.zip(y)).map { it.bimap(f, g) }
+
+      ls.equalUnderTheLaw(rs, EQ)
+    }
+
+  fun <F, A> Zip<F>.zippyness1(G: Gen<Kind<F, A>>, EQ: Eq<Kind<F, A>>) =
+    forAll(G) { x ->
+      x.zip(x).map { it.a }.equalUnderTheLaw(x, EQ)
+    }
+
+  fun <F, A> Zip<F>.zippyness2(G: Gen<Kind<F, A>>, EQ: Eq<Kind<F, A>>) =
+    forAll(G) { x ->
+      x.zip(x).map { it.b }.equalUnderTheLaw(x, EQ)
+    }
+
+  fun <F, A> Zip<F>.zippyness3(G: Gen<Kind<F, Tuple2<A, A>>>, EQ: Eq<Kind<F, Tuple2<A, A>>>) =
+    forAll(G) { x ->
+      (x.map { it.a }).zip(x.map { it.b }).equalUnderTheLaw(x, EQ)
+    }
+
+  fun <F, A> Zip<F>.distributivity1(G: Gen<Kind<F, A>>, EQ: Eq<Kind<F, Ior<Tuple2<A, A>, A>>>) =
+    forAll(G, G, G) { x, y, z ->
+      val ls = align(x.zip(y), z)
+      val rs = align(x, z).zip(align(y, z)).map { it.undistrThesePair() }
+
+      ls.equalUnderTheLaw(rs, EQ)
+    }
+
+  fun <F, A> Zip<F>.distributivity2(G: Gen<Kind<F, A>>, EQ: Eq<Kind<F, Ior<Tuple2<A, A>, Tuple2<A, A>>>>) =
+    forAll(G, G, G) { x, y, z ->
+      val ls = align(x, y).zip(z).map { it.distrPairThese() }
+      val rs = align(x.zip(z), y.zip(z))
+
+      ls.equalUnderTheLaw(rs, EQ)
+    }
+
+  fun <F, A> Zip<F>.distributivity3(G: Gen<Kind<F, A>>, EQ: Eq<Kind<F, Tuple2<Ior<A, A>, A>>>) =
+    forAll(G, G, G) { x, y, z ->
+      val ls = align(x, y).zip(z)
+      val rs = align(x.zip(z), y.zip(z)).map { it.undistrPairThese() }
+      ls.equalUnderTheLaw(rs, EQ)
+    }
+}
+
+private fun <A, B> Ior<A, B>.toLeft() =
+  fold({ this }, { this }, { l, _ -> Ior.Left(l) })
+
+private fun <A, B, C> Tuple2<Tuple2<A, B>, C>.assoc(): Tuple2<A, Tuple2<B, C>> =
+  this.a.a toT (this.a.b toT this.b)
+
+private fun <A, B, C> Tuple2<Ior<A, C>, Ior<B, C>>.undistrThesePair(): Ior<Tuple2<A, B>, C> =
+  when (val l = this.a) {
+    is Ior.Left -> {
+      when (val r = this.b) {
+        is Ior.Left -> Ior.Left(l.value toT r.value)
+        is Ior.Both -> Ior.Both(l.value toT r.leftValue, r.rightValue)
+        is Ior.Right -> Ior.Right(r.value)
+      }
+    }
+    is Ior.Both -> when (val r = this.b) {
+      is Ior.Left -> Ior.Both(l.leftValue toT r.value, l.rightValue)
+      is Ior.Both -> Ior.Both(l.leftValue toT r.leftValue, l.rightValue)
+      is Ior.Right -> Ior.Right(l.rightValue)
+    }
+    is Ior.Right -> Ior.Right(l.value)
+  }
+
+private fun <A, B, C> Tuple2<Ior<A, B>, C>.distrPairThese(): Ior<Tuple2<A, C>, Tuple2<B, C>> =
+  when (val l = this.a) {
+    is Ior.Left -> Ior.Left(l.value toT this.b)
+    is Ior.Right -> Ior.Right(l.value toT this.b)
+    is Ior.Both -> Ior.Both(l.leftValue toT this.b, l.rightValue toT this.b)
+  }
+
+private fun <A, B, C> Ior<Tuple2<A, C>, Tuple2<B, C>>.undistrPairThese(): Tuple2<Ior<A, B>, C> =
+  when (val e = this) {
+    is Ior.Left -> Ior.Left(e.value.a) toT e.value.b
+    is Ior.Both -> Ior.Both(e.leftValue.a, e.rightValue.a) toT e.leftValue.b
+    is Ior.Right -> Ior.Right(e.value.a) toT e.value.b
+  }
```

Comment:
This makes me want to add the `Distributive` typeclass to arrow... Either way, this is fine for now